pyro/infer/autoguide/guides.py

```diff
@@ -29,37 +29,47 @@ def model():
 from pyro.infer.enum import config_enumerate
 from pyro.nn import AutoRegressiveNN
 from pyro.ops.hessian import hessian
+from pyro.params import ConstrainedModule, ConstrainedParameter
 from pyro.poutine.util import prune_subsample_sites
 
 
-class AutoGuide(object):
+class AutoGuide(ConstrainedModule):
     """
     Base class for automatic guides.
 
-    Derived classes must implement the :meth:`__call__` method.
+    Derived classes must implement the :meth:`forward` method, with the
+    same ``*args, **kwargs`` as the base ``model``.
 
     Auto guides can be used individually or combined in an
     :class:`AutoGuideList` object.
 
     :param callable model: a pyro model
     :param str prefix: a prefix that will be prefixed to all param internal sites
+        as part of the module name.
     """
 
     def __init__(self, model, prefix="auto"):
+        super().__init__()
         self.master = None
         self.model = model
```

Comment:
I believe this will cause problems if `model` is an `nn.Module`, since e.g. `self.named_parameters` will iterate over model parameters. Do you think we should wrap it to bypass `nn.Module` behavior, e.g.
```py
def __init__(self, model, ...):
    self._model = (model,)
    ...

@property
def model(self):
    return self._model[0]
```

Reply:
Hmm..interesting. Let me add a test case and investigate further.

Reply:
It turned out that since the models are wrapped in an `InitMessenger` they fail the isinstance check for `nn.Module` and are therefore not registered as submodules for the guides. I am still using this for safety, and have added some cheap median tests for a model that is an `nn.Module` itself.